diff --git apache-couchdb-0.10.1/src/couchdb/couch_rep.erl apache-couchdb-0.10.1/src/couchdb/couch_rep.erl
index b57eea7..6e38d98 100644
--- apache-couchdb-0.10.1/src/couchdb/couch_rep.erl
+++ apache-couchdb-0.10.1/src/couchdb/couch_rep.erl
@@ -574,8 +574,7 @@ commit_to_both(Source, Target, RequiredSeq) ->
 ensure_full_commit(#http_db{} = Target) ->
     Req = Target#http_db{
         resource = "_ensure_full_commit",
-        method = post,
-        body = true
+        method = post
     },
     {ResultProps} = couch_rep_httpc:request(Req),
     true = proplists:get_value(<<"ok">>, ResultProps),
@@ -600,7 +599,6 @@ ensure_full_commit(#http_db{} = Source, RequiredSeq) ->
     Req = Source#http_db{
         resource = "_ensure_full_commit",
         method = post,
-        body = true,
         qs = [{seq, RequiredSeq}]
     },
     {ResultProps} = couch_rep_httpc:request(Req),
