diff --git apache-couchdb-0.10.1/etc/couchdb/default.ini.tpl.in apache-couchdb-0.10.1/etc/couchdb/default.ini.tpl.in
index 1a7643e..c02071a 100644
--- apache-couchdb-0.10.1/etc/couchdb/default.ini.tpl.in
+++ apache-couchdb-0.10.1/etc/couchdb/default.ini.tpl.in
@@ -93,3 +93,7 @@ _show = {couch_httpd_show, handle_doc_show_req}
 _list = {couch_httpd_show, handle_view_list_req}
 _info = {couch_httpd_db,   handle_design_info_req}
 _update = {couch_httpd_show, handle_doc_update_req}
+
+[replicator]
+max_http_sessions = 10
+max_http_pipeline_size = 10
\ No newline at end of file
diff --git apache-couchdb-0.10.1/src/couchdb/couch_db.hrl apache-couchdb-0.10.1/src/couchdb/couch_db.hrl
index ead7841..9cc8f47 100644
--- apache-couchdb-0.10.1/src/couchdb/couch_db.hrl
+++ apache-couchdb-0.10.1/src/couchdb/couch_db.hrl
@@ -242,7 +242,13 @@
     body = nil,
     options = [
         {response_format,binary},
-        {inactivity_timeout, 30000}
+        {inactivity_timeout, 30000},
+        {max_sessions, list_to_integer(
+            couch_config:get("replicator", "max_http_sessions", "10")
+        )},
+        {max_pipeline_size, list_to_integer(
+            couch_config:get("replicator", "max_http_pipeline_size", "10")
+        )}
     ],
     retries = 10,
     pause = 500,
